### PR TITLE
Optimize reactions table

### PR DESCRIPTION
**Pull request by Pomyk**
#### Summary
Change reactions primary key to (PostId, UserId, EmojiName) so fetching
reactions for post will use primary key lookup instead of table scan.

### CODE REVIEW

**Comment by grundleborg** [comment]
Thanks very much for this PR @Pomyk . Will this migration require the rewriting of the primary key index for the Reactions table? If so, will it take a long time if, for example, you have 100,0000,000 reactions in your database?

I'm wondering if this is something we should hold back on until we have an asynchronous way to make such schema changes. Or alternatively, make the change for *new* installs only for now, and defer changing old installs until we have a way to do that without downtime on upgrading the Mattermost version. Or perhaps considering a way system admins can opt in to such changes. The more I think about this, in fact, the more I realise that there's a product management question around how best to do such changes once we have a clear list of what is and is not possible from a *technical implementation* point of view.

**Comment by Pomyk** [comment]
I think if anyone had 100 mln reactions they would notice slow queries as it does a table scan on this table for each post with reactions.

**Comment by grundleborg** [comment]
I've (finally) had a chance to do some interim testing on this. I've verified the query plan around that query on a medium sized database using MySQL (around half a million rows in the `Reactions` table), and indeed, the plan changes from a full table scan to a primary key index lookup. In my test the drop/create of the index took 15 seconds to run, which isn't too long.

Next testing plans:
1. Manually look through all the `Reactions` table queries, and check plans before and after the change, just to be 100% sure there's no disadvantage of this change.
2. Do a full load test against this branch and check none of the API endpoints shows any slow down as a result of the change.
3. Find out what a really big actual real-world installed Mattermost database looks like, in terms of the size of the `Reactions` table, and load up a test instance to that scale, and test before and after, as well as the time to recreate the primary key. Based on this, decide whether to merge it as is, or whether this needs to be held back for the asynchronous-schema-changes branch to be merged first.

**Comment by Pomyk** [comment]
I didn't find any query that would scan only for UserId.

**Comment by jasonblais** [comment]
@grundleborg Wondering if there are any follow-up tickets for the above testing plan?

**Comment by Pomyk** [comment]
Maybe you could merge the fix in create table first so at least new installations get better performance?

**Comment by grundleborg** [comment]
So I've checked this on Postgres and all seems to be good. And the hoped for performance gains seem to be materialised with the new index, so that's looking good too.

Just talking with @wiersgallak as to whether we just push the whole thing in for 5.22 now. Will revert back here when we reach a conclusion.

File: store/sqlstore/supplier.go [hunk 1]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+		AND c.conrelid = '` + strings.ToLower(tableName) + `'::REGCLASS`
+		currentPrimaryKey, err = ss.GetMaster().SelectStr(sql)
+	} else if ss.DriverName() == model.DATABASE_DRIVER_SQLITE {
+		// SQLite doesn't support altering primary key
+		return true
+	}
+	if err != nil {
+		mlog.Critical("Failed to get current primary key", mlog.String("table", tableName), mlog.Err(err))
+		time.Sleep(time.Second)
+		os.Exit(EXIT_GENERIC_FAILURE)
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: I think we can return a more specific error code to be consistent with the rest of the code.

File: store/sqlstore/supplier.go [hunk 2]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+	if ss.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		alter = "ALTER TABLE " + tableName + " DROP PRIMARY KEY, ADD PRIMARY KEY (" + primaryKey + ")"
+	} else if ss.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		alter = "ALTER TABLE " + tableName + " DROP CONSTRAINT " + strings.ToLower(tableName) + "_pkey, ADD PRIMARY KEY (" + strings.ToLower(primaryKey) + ")"
+	}
+	_, err = ss.GetMaster().ExecNoTimeout(alter)
+	if err != nil {
+		mlog.Critical("Failed to alter primary key", mlog.String("table", tableName), mlog.Err(err))
+		time.Sleep(time.Second)
+		os.Exit(EXIT_GENERIC_FAILURE)
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: Here also I would use a less generic error code.

File: store/sqlstore/supplier.go [hunk 3]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
 	return true
 }
 
+func (ss *SqlSupplier) AlterPrimaryKey(tableName string, columnNames []string) bool {
+	var currentPrimaryKey string
+	var err error
+	// get the current primary key as a comma separated list of columns
+	if ss.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		sql := `
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: I'd probably call this var `query`.

File: store/sqlstore/supplier.go [hunk 4]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+			information_schema.statistics
+		WHERE
+			table_schema = DATABASE()
+		AND table_name = ?
+		AND index_name = 'PRIMARY'
+		GROUP BY
+			index_name`
+		currentPrimaryKey, err = ss.GetMaster().SelectStr(sql, tableName)
+	} else if ss.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		sql := `
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: Same here.

File: store/sqlstore/supplier.go [hunk 5]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+		os.Exit(EXIT_GENERIC_FAILURE)
+		return false
+	}
+
+	primaryKey := strings.Join(columnNames, ",")
+	if strings.EqualFold(currentPrimaryKey, primaryKey) {
+		return false
+	}
+	// alter primary key
+	var alter string
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: Let's call it `alterQuery`.

File: store/sqlstore/upgrade.go [hunk 6]
```diff
@@ -753,6 +753,7 @@ func upgradeDatabaseToVersion520(sqlStore SqlStore) {
 	// if shouldPerformUpgrade(sqlStore, VERSION_5_19_0, VERSION_5_20_0) {
 
 	sqlStore.CreateColumnIfNotExistsNoDefault("Bots", "LastIconUpdate", "bigint", "bigint")
+	sqlStore.AlterPrimaryKey("Reactions", []string{"PostId", "UserId", "EmojiName"})
```
Comment (streamer45) [inline store/sqlstore/upgrade.go]: Just a reminder to move this to 5.22

File: store/sqlstore/supplier.go [hunk 7]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+		alter = "ALTER TABLE " + tableName + " DROP PRIMARY KEY, ADD PRIMARY KEY (" + primaryKey + ")"
+	} else if ss.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		alter = "ALTER TABLE " + tableName + " DROP CONSTRAINT " + strings.ToLower(tableName) + "_pkey, ADD PRIMARY KEY (" + strings.ToLower(primaryKey) + ")"
+	}
+	_, err = ss.GetMaster().ExecNoTimeout(alter)
+	if err != nil {
+		mlog.Critical("Failed to alter primary key", mlog.String("table", tableName), mlog.Err(err))
+		time.Sleep(time.Second)
+		os.Exit(EXIT_GENERIC_FAILURE)
+		return false
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: No need to return here since we exit the program.

File: store/sqlstore/supplier.go [hunk 8]
```diff
@@ -743,6 +743,68 @@ func (ss *SqlSupplier) AlterColumnDefaultIfExists(tableName string, columnName s
...
+		currentPrimaryKey, err = ss.GetMaster().SelectStr(sql)
+	} else if ss.DriverName() == model.DATABASE_DRIVER_SQLITE {
+		// SQLite doesn't support altering primary key
+		return true
+	}
+	if err != nil {
+		mlog.Critical("Failed to get current primary key", mlog.String("table", tableName), mlog.Err(err))
+		time.Sleep(time.Second)
+		os.Exit(EXIT_GENERIC_FAILURE)
+		return false
```
Comment (streamer45) [inline store/sqlstore/supplier.go]: No need to return here since we exit the program.

**Comment by streamer45** [comment]
Looks like the schema diff is failing.
We should probably update these:

https://github.com/mattermost/mattermost-server/blob/1e53fe85ad42653f1cbdeb9a1d4447bc3b6d7b3d/scripts/mattermost-postgresql-5.0.sql#L1082-L1083

**Pull request merged**